storage: TestMVCCHistories failed

storage.TestMVCCHistories [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11806630?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11806630?buildTab=artifacts#/) on master @ [985662236d7bf273b93a7b5e32def8e2d1043640](https://github.com/cockroachdb/cockroach/commits/985662236d7bf273b93a7b5e32def8e2d1043640):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x177c263]

```
Stack: 

```
goroutine 25982 [running]:
testing.tRunner.func1.2({0x2b6e800, 0x48498b0})
	GOROOT/src/testing/testing.go:1526 +0x372
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1529 +0x650
panic({0x2b6e800, 0x48498b0})
	GOROOT/src/runtime/panic.go:890 +0x263
github.com/cockroachdb/datadriven.runDirective.func1.1()
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:334 +0x165
panic({0x2b6e800, 0x48498b0})
	GOROOT/src/runtime/panic.go:890 +0x263
github.com/cockroachdb/pebble/internal/base.(*LazyValue).fetchValue(0xc000ff47e8, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/internal/base/external/com_github_cockroachdb_pebble/internal/base/lazy_value.go:212 +0x143
github.com/cockroachdb/pebble/internal/base.(*LazyValue).Value(0xc000ff47e8, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/internal/base/external/com_github_cockroachdb_pebble/internal/base/lazy_value.go:204 +0x69
github.com/cockroachdb/pebble.(*Iterator).ValueAndErr(0xc000ff4600)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:2138 +0x71
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).UnsafeValue(0xc000b9c170)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:522 +0xd9
github.com/cockroachdb/cockroach/pkg/storage.assertMVCCIteratorInvariants({0x38a1100?, 0xc000b9c170})
	github.com/cockroachdb/cockroach/pkg/storage/engine.go:1944 +0x9c2
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).assertMVCCInvariants(0xc000b9c170)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:980 +0x52
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Valid(0xc000b9c170)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:419 +0x19a
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).tryDecodeKey(0xc000f5a200)
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:648 +0x67
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekLT(0xc000f5a200, {{0xc00087c068, 0x1, 0x8}, {0x1a13b8600, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1154 +0x48a
github.com/cockroachdb/cockroach/pkg/storage.(*unsafeMVCCIterator).SeekLT(0xc0005f57a0, {{0xc00087c068, 0x1, 0x8}, {0x1a13b8600, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1571 +0xbc
github.com/cockroachdb/cockroach/pkg/storage_test.(*metamorphicMVCCIterator).SeekLT(0xc00011c4d8, {{0xc00087c068, 0x1, 0x8}, {0x1a13b8600, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_metamorphic_iterator_test.go:359 +0xf2
github.com/cockroachdb/cockroach/pkg/storage_test.cmdIterSeekLT(0xc000dc2200?)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:1931 +0x14a
github.com/cockroachdb/cockroach/pkg/storage_test.TestMVCCHistories.func1.5(0xc00187f1e0, 0xc000a5e028)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:632 +0x1690
github.com/cockroachdb/datadriven.RunTest.func1({0x389c718?, 0xc00187f1e0}, 0x23952ec?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118 +0x6b
github.com/cockroachdb/datadriven.runDirective.func1({0x389c718, 0xc00187f1e0}, 0xc000a5e028, 0xc000355890)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 +0x102
github.com/cockroachdb/datadriven.runDirective({0x389c718, 0xc00187f1e0}, 0xc000a5e000, 0x0?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342 +0x9a
github.com/cockroachdb/datadriven.runDirectiveOrSubTest({0x389c718, 0xc00187f1e0}, 0xc00187f1e0?, {0x0, 0x0}, 0xc00011c410?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211 +0xab
github.com/cockroachdb/datadriven.runTestInternal({0x389c718, 0xc00187f1e0}, {0xc00006a140, 0x139}, {0x386cc20, 0xc00011c410}, 0x496f77?, 0xec?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:184 +0xf2
github.com/cockroachdb/datadriven.RunTestAny({0x389c718, 0xc00187f1e0}, {0xc00006a140, 0x139}, 0x50?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 +0x3d2
github.com/cockroachdb/datadriven.RunTest(0xc00187f1e0, {0xc00006a140, 0x139}, 0xc000438780)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 +0xc5
github.com/cockroachdb/cockroach/pkg/storage_test.TestMVCCHistories.func1(0xc00187f1e0, {0xc00006a140, 0x139})
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:429 +0xd8c
github.com/cockroachdb/datadriven.Walk.func1({0x389c718?, 0xc00187f1e0}, {0xc00006a140, 0x139})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 +0x76
github.com/cockroachdb/datadriven.WalkAny({0x389c718, 0xc00187f1e0}, {0xc00006a140, 0x139}, 0xc0065726d0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 +0x298
github.com/cockroachdb/datadriven.WalkAny.func1({0x389c718, 0xc00187f1e0})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 +0xc8
github.com/cockroachdb/datadriven.subTest.func1(0x0?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:27 +0x42
testing.tRunner(0xc00187f1e0, 0xc000c1a0e0)
	GOROOT/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1629 +0x806
```
<details><summary>Log preceding fatal error</summary>
<p>

```
        ----
        iter_seek_ge: "j"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {h-k}/[1.000000000,0=/<empty>] !
        iter_next: "j"/7.000000000,0=/BYTES/j7 {h-k}/[1.000000000,0=/<empty>]
        iter_prev: "j"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {h-k}/[1.000000000,0=/<empty>]
        iter_prev: "h"/3.000000000,0=/BYTES/h3 {h-k}/[1.000000000,0=/<empty>]
        iter_prev: "h"/4.000000000,0=/<empty> {h-k}/[1.000000000,0=/<empty>]
        iter_seek_ge: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_next: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: {c-d}/[5.000000000,0=/<empty> 3.000000000,0=/<empty> 1.000000000,0=/<empty>] !
    mvcc_history_metamorphic_iterator_test.go:286: metamorphicIterator: carried out 9 actions
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1995/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_iter:650:
        run [1 args]
        iter_new types=pointsAndRanges
        iter_seek_ge k=d
        iter_prev
        iter_seek_ge k=d ts=8
        iter_prev
        iter_seek_ge k=d ts=7
        iter_prev
        iter_seek_ge k=d ts=5
        iter_prev
        iter_seek_ge k=d ts=4
        iter_prev
        ----
        iter_seek_ge: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_prev: {c-d}/[5.000000000,0=/<empty> 3.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_seek_ge: {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_prev: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_seek_ge: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_seek_ge: {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_seek_ge: "d"/4.000000000,0=/BYTES/d4 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
    mvcc_history_metamorphic_iterator_test.go:286: metamorphicIterator: carried out 10 actions
    datadriven.go:333: 
        panic during /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/1995/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_iter:674:
        iter_new types=pointsAndRanges
        iter_seek_lt k=e
        iter_next
        iter_seek_lt k=d ts=4
        iter_next
        iter_seek_lt k=d ts=7
        iter_next
        iter_seek_lt k=d
        iter_next
    mvcc_history_metamorphic_iterator_test.go:286: metamorphicIterator: carried out 4 actions
    --- FAIL: TestMVCCHistories/range_tombstone_iter (0.83s)
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMVCCHistories.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-31604

storage.TestMVCCHistories [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11824254?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11824254?buildTab=artifacts#/) on master @ [d414347adc48c16951ed1092d1fcc47ef6f2adb0](https://github.com/cockroachdb/cockroach/commits/d414347adc48c16951ed1092d1fcc47ef6f2adb0):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x177c263]

```
Stack: 

```
goroutine 25529 [running]:
testing.tRunner.func1.2({0x2b6ec40, 0x484a8b0})
	GOROOT/src/testing/testing.go:1526 +0x372
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1529 +0x650
panic({0x2b6ec40, 0x484a8b0})
	GOROOT/src/runtime/panic.go:890 +0x263
github.com/cockroachdb/datadriven.runDirective.func1.1()
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:334 +0x165
panic({0x2b6ec40, 0x484a8b0})
	GOROOT/src/runtime/panic.go:890 +0x263
github.com/cockroachdb/pebble/internal/base.(*LazyValue).fetchValue(0xc000d6ade8, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/internal/base/external/com_github_cockroachdb_pebble/internal/base/lazy_value.go:212 +0x143
github.com/cockroachdb/pebble/internal/base.(*LazyValue).Value(0xc000d6ade8, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/internal/base/external/com_github_cockroachdb_pebble/internal/base/lazy_value.go:204 +0x69
github.com/cockroachdb/pebble.(*Iterator).ValueAndErr(0xc000d6ac00)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:2138 +0x71
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).UnsafeValue(0xc000879370)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:522 +0xd9
github.com/cockroachdb/cockroach/pkg/storage.assertMVCCIteratorInvariants({0x38a1640?, 0xc000879370})
	github.com/cockroachdb/cockroach/pkg/storage/engine.go:1944 +0x9c2
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).assertMVCCInvariants(0xc000879370)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:980 +0x52
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Valid(0xc000879370)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:419 +0x19a
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).tryDecodeKey(0xc000cd8600)
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:648 +0x67
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekLT(0xc000cd8600, {{0xc00028a625, 0x1, 0xb}, {0x0, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1154 +0x48a
github.com/cockroachdb/cockroach/pkg/storage.(*unsafeMVCCIterator).SeekLT(0xc000eca7e0, {{0xc00028a625, 0x1, 0xb}, {0x0, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1571 +0xbc
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).iterSeekReverse(0xc000879800, {{0xc00028a625, 0x1, 0xb}, {0x0, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1664 +0xf1
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).prevKey(0xc000879800, {0xc00028a625, 0x1, 0xb})
	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1145 +0x2a6
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).advanceKeyReverse(...)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:1157
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).advance(0xc000879800)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:633 +0x3e7
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleMVCCScanner).scan(0xc000879800, {0x3883138, 0xc0001b4008})
	github.com/cockroachdb/cockroach/pkg/storage/pebble_mvcc_scanner.go:661 +0x18c
github.com/cockroachdb/cockroach/pkg/storage.mvccScanToBytes({_, _}, {_, _}, {_, _, _}, {0xc001a12db8, 0x1, 0x8}, ...)
	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:3899 +0x42a
github.com/cockroachdb/cockroach/pkg/storage.mvccScanToKvs({_, _}, {_, _}, {_, _, _}, {0xc001a12db8, 0x1, 0x8}, ...)
	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:3952 +0x1b0
github.com/cockroachdb/cockroach/pkg/storage.MVCCScan({_, _}, {_, _}, {_, _, _}, {0xc001a12db8, 0x1, 0x8}, ...)
	github.com/cockroachdb/cockroach/pkg/storage/mvcc.go:4183 +0x350
github.com/cockroachdb/cockroach/pkg/storage_test.cmdScan.func1({0x388eb78, 0xc000879200})
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:1719 +0x205
github.com/cockroachdb/cockroach/pkg/storage_test.(*evalCtx).withReader(0xc000272100?, 0xc000dbc6e0)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:2453 +0xc7
github.com/cockroachdb/cockroach/pkg/storage_test.cmdScan(0xc000272100)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:1718 +0xf65
github.com/cockroachdb/cockroach/pkg/storage_test.TestMVCCHistories.func1.5(0xc000022680, 0xc000714148)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:632 +0x1690
github.com/cockroachdb/datadriven.RunTest.func1({0x389cc58?, 0xc000022680}, 0x239564c?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118 +0x6b
github.com/cockroachdb/datadriven.runDirective.func1({0x389cc58, 0xc000022680}, 0xc000714148, 0xc005306280)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 +0x102
github.com/cockroachdb/datadriven.runDirective({0x389cc58, 0xc000022680}, 0xc000714120, 0x0?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342 +0x9a
github.com/cockroachdb/datadriven.runDirectiveOrSubTest({0x389cc58, 0xc000022680}, 0xc000022680?, {0x0, 0x0}, 0xc0001a8068?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211 +0xab
github.com/cockroachdb/datadriven.runTestInternal({0x389cc58, 0xc000022680}, {0xc000cd2000, 0x142}, {0x386d160, 0xc0001a8068}, 0x496f77?, 0x8c?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:184 +0xf2
github.com/cockroachdb/datadriven.RunTestAny({0x389cc58, 0xc000022680}, {0xc000cd2000, 0x142}, 0x50?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 +0x3d2
github.com/cockroachdb/datadriven.RunTest(0xc000022680, {0xc000cd2000, 0x142}, 0xc000c5a410)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 +0xc5
github.com/cockroachdb/cockroach/pkg/storage_test.TestMVCCHistories.func1(0xc000022680, {0xc000cd2000, 0x142})
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:429 +0xd8c
github.com/cockroachdb/datadriven.Walk.func1({0x389cc58?, 0xc000022680}, {0xc000cd2000, 0x142})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 +0x76
github.com/cockroachdb/datadriven.WalkAny({0x389cc58, 0xc000022680}, {0xc000cd2000, 0x142}, 0xc005306dd0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 +0x298
github.com/cockroachdb/datadriven.WalkAny.func1({0x389cc58, 0xc000022680})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 +0xc8
github.com/cockroachdb/datadriven.subTest.func1(0x0?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:27 +0x42
testing.tRunner(0xc000022680, 0xc000682440)
	GOROOT/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1629 +0x806
```
<details><summary>Log preceding fatal error</summary>
<p>

```
        scan: "h" -> /BYTES/h3 @3.000000000,0
        scan: "k" -> /BYTES/k5 @5.000000000,0
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_scans_complex:212:
        run [1 args]
        scan k=a end=z ts=6 tombstones inconsistent
        ----
        scan: "a" -> /<empty> @4.000000000,0
        scan: "b" -> /<empty> @4.000000000,0
        scan: "d" -> /<empty> @5.000000000,0
        scan: "e" -> /<empty> @5.000000000,0
        scan: "f" -> /BYTES/f6 @6.000000000,0
        scan: "g" -> /<empty> @5.000000000,0
        scan: "h" -> /BYTES/h3 @3.000000000,0
        scan: "k" -> /BYTES/k5 @5.000000000,0
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_scans_complex:224:
        run [1 args]
        scan k=a end=z ts=7 tombstones inconsistent
        ----
        scan: intent "d" {id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0}
        scan: intent "f" {id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0}
        scan: intent "j" {id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0}
        scan: "a" -> /<empty> @4.000000000,0
        scan: "b" -> /<empty> @4.000000000,0
        scan: "d" -> /<empty> @5.000000000,0
        scan: "e" -> /<empty> @5.000000000,0
        scan: "f" -> /BYTES/f6 @6.000000000,0
        scan: "g" -> /<empty> @5.000000000,0
        scan: "h" -> /BYTES/h3 @3.000000000,0
        scan: "k" -> /BYTES/k5 @5.000000000,0
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_scans_complex:240:
        run [1 args]
        scan k=a end=z ts=1 reverse
        ----
        scan: "a"-"z" -> <no data>
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_scans_complex:245:
        run [1 args]
        scan k=a end=z ts=2 reverse
        ----
        scan: "g" -> /BYTES/g2 @2.000000000,0
        scan: "f" -> /BYTES/f2 @2.000000000,0
        scan: "d" -> /BYTES/d2 @2.000000000,0
        scan: "a" -> /BYTES/a2 @2.000000000,0
    datadriven.go:333: 
        panic during /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_scans_complex:253:
        scan k=a end=z ts=3 reverse
    --- FAIL: TestMVCCHistories/range_tombstone_scans_complex (0.24s)
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMVCCHistories.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

storage.TestMVCCHistories [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11840742?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/11840742?buildTab=artifacts#/) on master @ [33623e3ee420174a4fd3226d1284b03f0e3caaac](https://github.com/cockroachdb/cockroach/commits/33623e3ee420174a4fd3226d1284b03f0e3caaac):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x177c0c3]

```
Stack: 

```
goroutine 26325 [running]:
testing.tRunner.func1.2({0x2b71700, 0x484e8b0})
	GOROOT/src/testing/testing.go:1526 +0x372
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1529 +0x650
panic({0x2b71700, 0x484e8b0})
	GOROOT/src/runtime/panic.go:890 +0x263
github.com/cockroachdb/datadriven.runDirective.func1.1()
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:334 +0x165
panic({0x2b71700, 0x484e8b0})
	GOROOT/src/runtime/panic.go:890 +0x263
github.com/cockroachdb/pebble/internal/base.(*LazyValue).fetchValue(0xc000d9c1e8, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/internal/base/external/com_github_cockroachdb_pebble/internal/base/lazy_value.go:212 +0x143
github.com/cockroachdb/pebble/internal/base.(*LazyValue).Value(0xc000d9c1e8, {0x0, 0x0, 0x0})
	github.com/cockroachdb/pebble/internal/base/external/com_github_cockroachdb_pebble/internal/base/lazy_value.go:204 +0x69
github.com/cockroachdb/pebble.(*Iterator).ValueAndErr(0xc000d9c000)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/iterator.go:2141 +0x71
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).UnsafeValue(0xc000ea0170)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:522 +0xd9
github.com/cockroachdb/cockroach/pkg/storage.assertMVCCIteratorInvariants({0x38a48c0?, 0xc000ea0170})
	github.com/cockroachdb/cockroach/pkg/storage/engine.go:1949 +0x9c2
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).assertMVCCInvariants(0xc000ea0170)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:980 +0x52
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Valid(0xc000ea0170)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:419 +0x19a
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).tryDecodeKey(0xc0005f8900)
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:648 +0x67
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).SeekLT(0xc0005f8900, {{0xc000d05208, 0x1, 0x8}, {0x1a13b8600, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1154 +0x48a
github.com/cockroachdb/cockroach/pkg/storage.(*unsafeMVCCIterator).SeekLT(0xc0050564e0, {{0xc000d05208, 0x1, 0x8}, {0x1a13b8600, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1571 +0xbc
github.com/cockroachdb/cockroach/pkg/storage_test.(*metamorphicMVCCIterator).SeekLT(0xc00061e028, {{0xc000d05208, 0x1, 0x8}, {0x1a13b8600, 0x0, 0x0}})
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_metamorphic_iterator_test.go:359 +0xf2
github.com/cockroachdb/cockroach/pkg/storage_test.cmdIterSeekLT(0xc0005f8c00?)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:1931 +0x14a
github.com/cockroachdb/cockroach/pkg/storage_test.TestMVCCHistories.func1.5(0xc0008221a0, 0xc000f6a5c8)
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:632 +0x1690
github.com/cockroachdb/datadriven.RunTest.func1({0x389fed8?, 0xc0008221a0}, 0x239542c?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118 +0x6b
github.com/cockroachdb/datadriven.runDirective.func1({0x389fed8, 0xc0008221a0}, 0xc000f6a5c8, 0xc00063c960)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 +0x102
github.com/cockroachdb/datadriven.runDirective({0x389fed8, 0xc0008221a0}, 0xc000f6a5a0, 0x0?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342 +0x9a
github.com/cockroachdb/datadriven.runDirectiveOrSubTest({0x389fed8, 0xc0008221a0}, 0xc0008221a0?, {0x0, 0x0}, 0xc000628058?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211 +0xab
github.com/cockroachdb/datadriven.runTestInternal({0x389fed8, 0xc0008221a0}, {0xc000ad2140, 0x139}, {0x38703e0, 0xc000628058}, 0x496f77?, 0xac?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:184 +0xf2
github.com/cockroachdb/datadriven.RunTestAny({0x389fed8, 0xc0008221a0}, {0xc000ad2140, 0x139}, 0x50?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 +0x3d2
github.com/cockroachdb/datadriven.RunTest(0xc0008221a0, {0xc000ad2140, 0x139}, 0xc0005bebe0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 +0xc5
github.com/cockroachdb/cockroach/pkg/storage_test.TestMVCCHistories.func1(0xc0008221a0, {0xc000ad2140, 0x139})
	github.com/cockroachdb/cockroach/pkg/storage_test/pkg/storage/mvcc_history_test.go:429 +0xd8c
github.com/cockroachdb/datadriven.Walk.func1({0x389fed8?, 0xc0008221a0}, {0xc000ad2140, 0x139})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 +0x76
github.com/cockroachdb/datadriven.WalkAny({0x389fed8, 0xc0008221a0}, {0xc000ad2140, 0x139}, 0xc0001a2a70)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 +0x298
github.com/cockroachdb/datadriven.WalkAny.func1({0x389fed8, 0xc0008221a0})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 +0xc8
github.com/cockroachdb/datadriven.subTest.func1(0x0?)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:27 +0x42
testing.tRunner(0xc0008221a0, 0xc00063c050)
	GOROOT/src/testing/testing.go:1576 +0x217
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:1629 +0x806
```
<details><summary>Log preceding fatal error</summary>
<p>

```
        ----
        iter_seek_ge: "j"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {h-k}/[1.000000000,0=/<empty>] !
        iter_next: "j"/7.000000000,0=/BYTES/j7 {h-k}/[1.000000000,0=/<empty>]
        iter_prev: "j"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {h-k}/[1.000000000,0=/<empty>]
        iter_prev: "h"/3.000000000,0=/BYTES/h3 {h-k}/[1.000000000,0=/<empty>]
        iter_prev: "h"/4.000000000,0=/<empty> {h-k}/[1.000000000,0=/<empty>]
        iter_seek_ge: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_next: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: {c-d}/[5.000000000,0=/<empty> 3.000000000,0=/<empty> 1.000000000,0=/<empty>] !
    mvcc_history_metamorphic_iterator_test.go:286: metamorphicIterator: carried out 9 actions
    datadriven.go:144: 
        /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_iter:650:
        run [1 args]
        iter_new types=pointsAndRanges
        iter_seek_ge k=d
        iter_prev
        iter_seek_ge k=d ts=8
        iter_prev
        iter_seek_ge k=d ts=7
        iter_prev
        iter_seek_ge k=d ts=5
        iter_prev
        iter_seek_ge k=d ts=4
        iter_prev
        ----
        iter_seek_ge: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_prev: {c-d}/[5.000000000,0=/<empty> 3.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_seek_ge: {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>] !
        iter_prev: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_seek_ge: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/0,0=txn={id=00000001 key=/Min iso=Serializable pri=0.00000000 epo=0 ts=7.000000000,0 min=0,0 seq=0} ts=7.000000000,0 del=false klen=12 vlen=7 mergeTs=<nil> txnDidNotUpdateMeta=true {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_seek_ge: {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_seek_ge: "d"/4.000000000,0=/BYTES/d4 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
        iter_prev: "d"/7.000000000,0=/BYTES/d7 {d-f}/[5.000000000,0=/<empty> 1.000000000,0=/<empty>]
    mvcc_history_metamorphic_iterator_test.go:286: metamorphicIterator: carried out 10 actions
    datadriven.go:333: 
        panic during /home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/3300/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/storage/storage_test_/storage_test.runfiles/com_github_cockroachdb_cockroach/pkg/storage/testdata/mvcc_histories/range_tombstone_iter:674:
        iter_new types=pointsAndRanges
        iter_seek_lt k=e
        iter_next
        iter_seek_lt k=d ts=4
        iter_next
        iter_seek_lt k=d ts=7
        iter_next
        iter_seek_lt k=d
        iter_next
    mvcc_history_metamorphic_iterator_test.go:286: metamorphicIterator: carried out 4 actions
    --- FAIL: TestMVCCHistories/range_tombstone_iter (0.76s)
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMVCCHistories.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/storage/engine.go
+++ b/pkg/storage/engine.go
@@ -1945,17 +1945,21 @@ func assertMVCCIteratorInvariants(iter MVCCIterator) error {
 		return errors.AssertionFailedf("unknown type for engine key %s", engineKey)
 	}
 
-	// Value must equal UnsafeValue.
-	u, err := iter.UnsafeValue()
-	if err != nil {
-		return err
-	}
-	v, err := iter.Value()
-	if err != nil {
-		return err
-	}
-	if !bytes.Equal(v, u) {
-		return errors.AssertionFailedf("Value %x does not match UnsafeValue %x at %s", v, u, key)
+	// If the iterator position has a point key, Value must equal UnsafeValue.
+	// NB: It's only valid to read an iterator's Value if the iterator is
+	// positioned at a point key.
+	if hasPoint, _ := iter.HasPointAndRange(); hasPoint {
+		u, err := iter.UnsafeValue()
+		if err != nil {
+			return err
+		}
+		v, err := iter.Value()
+		if err != nil {
+			return err
+		}
+		if !bytes.Equal(v, u) {
+			return errors.AssertionFailedf("Value %x does not match UnsafeValue %x at %s", v, u, key)
+		}
 	}
 
 	// For prefix iterators, any range keys must be point-sized. We've already

